### PR TITLE
Add Lets Encrypt sample and article

### LINKED ISSUE #110: Support for Lets Encrypt for TLS termination
### What should we add or change to make your life better?

Most reverse proxies out there support Lets Encrypt. @natemcmaster runs https://github.com/natemcmaster/LetsEncrypt which is trying to add LetsEncrypt support to ASP.NET Core. Please help him finish, so it can also be used on this project.

### Why is this important to you?

A reverse proxy without TLS termination is not very useful. Lets Encrypt is a great choice for TLS, being free, secure and automated.

### PR DESCRIPTION
**Pull request by greenEkatherine**
Closes #110

### CODE REVIEW

File: samples/ReverseProxy.LetsEncrypt.Sample/appsettings.json [hunk 1]
```diff
@@ -0,0 +1,50 @@
+{
+  "Kestrel": {
+    "Endpoints": {
+      "Http": {
+        "Url": "http://localhost:80"
+      },
+      "Https": {
+        "Url": "https://localhost:443"
```
Comment (Tratcher) [inline samples/ReverseProxy.LetsEncrypt.Sample/appsettings.json:8]: Localhost won't work if this site is running publicly on the internet.
```suggestion
        "Url": "http://*:80"
      },
      "Https": {
        "Url": "https://*:443"
```

File: samples/ReverseProxy.LetsEncrypt.Sample/Program.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        public static void Main(string[] args)
+        {
+            // Create a Kestrel web server, and tell it to use the Startup class
+            // for the service configuration
+            var myHostBuilder = Host.CreateDefaultBuilder(args);
+            myHostBuilder.ConfigureWebHostDefaults(webHostBuilder =>
+            {
+                webHostBuilder.ConfigureKestrel(kestrel =>
+                {
+                    kestrel.ListenAnyIP(443, portOptions =>
```
Comment (Tratcher) [inline samples/ReverseProxy.LetsEncrypt.Sample/Program.cs:20]: This is redundant with the appsettings.json endpoints, right? If so, remove it, we should only show one or the other in the sample.

File: docs/docfx/articles/lets-encrypt.md [hunk 3]
```diff
@@ -0,0 +1,93 @@
+# Lets Encrypt
+
+## Introduction
+YARP can support a certificate authority [Lets Encrypt](https://letsencrypt.org/) by using the API of another ASP.NET Core project [LettuceEncrypt](https://github.com/natemcmaster/LettuceEncrypt). It allows to set up TLS between the client and YARP and then use HTTP communication to the backend.
```
Comment (Tratcher) [inline docs/docfx/articles/lets-encrypt.md:4]:
```suggestion
YARP can support the certificate authority [Lets Encrypt](https://letsencrypt.org/) by using the API of another ASP.NET Core project [LettuceEncrypt](https://github.com/natemcmaster/LettuceEncrypt). It allows you to set up a TLS between the client and YARP and then use HTTP communication to the backend.
```

File: docs/docfx/articles/lets-encrypt.md [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+## Requirements
+
+LettuceEncrypt package should be added into project:
```
Comment (Tratcher) [inline docs/docfx/articles/lets-encrypt.md:8]:
```suggestion
Add the LettuceEncrypt package dependency:
```

File: docs/docfx/articles/lets-encrypt.md [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+```
+
+## Configuration
+There are required options for LettuceEncrypt that should be set, see the example of `appsettings.json`:
+
+```JSON
+{
+  // Base URLs the server listens on, must be configured independently of the routes below.
+  // Can also be configured via Kestrel/Endpoints, see https://docs.microsoft.com/en-us/aspnet/core/fundamentals/servers/kestrel/endpoints
+  "Urls": "http://localhost:80;https://localhost:443",
```
Comment (Tratcher) [inline docs/docfx/articles/lets-encrypt.md:20]:
```suggestion
  "Urls": "http://*:80;https://*:443",
```

File: docs/docfx/articles/lets-encrypt.md [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        services.AddLettuceEncrypt();
+    }
+}
+```
+
+For more options (i.e. saving certificates) see examples in [LettuceEncrypt doc](ttps://github.com/natemcmaster/LettuceEncrypt).
+
+## Middleware
+
+If your project is explicitly using kestrel options to configure IP addresses, ports, or HTTPS settings, you will also need to call `UseLettuceEncrypt`. This is required to make Lettuce Encrypt work.
```
Comment (Tratcher) [inline docs/docfx/articles/lets-encrypt.md:72]:
```suggestion
If your project is explicitly using kestrel options to configure IP addresses, ports, or HTTPS settings, you will also need to call `UseLettuceEncrypt`.
```

File: samples/ReverseProxy.LetsEncrypt.Sample/Program.cs (hunk 2 above)
Comment (greenEkatherine) [reply samples/ReverseProxy.LetsEncrypt.Sample/Program.cs]: Removed

File: docs/docfx/articles/lets-encrypt.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
+# Lets Encrypt
+
+## Introduction
+YARP can support the certificate authority [Lets Encrypt](https://letsencrypt.org/) by using the API of another ASP.NET Core project [LettuceEncrypt](https://github.com/natemcmaster/LettuceEncrypt). It allows you to set up a TLS between the client and YARP and then use HTTP communication to the backend.
```
Comment (MihaZupan) [inline docs/docfx/articles/lets-encrypt.md:4]:
```suggestion
YARP can support the certificate authority [Lets Encrypt](https://letsencrypt.org/) by using the API of another ASP.NET Core project [LettuceEncrypt](https://github.com/natemcmaster/LettuceEncrypt). It allows you to set up TLS between the client and YARP with minimal configuration.
```

Do we want to call out using http to the backend here? Technically it's not related to what happens between the client and the proxy.

File: docs/docfx/articles/lets-encrypt.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    // If you don't set this in config, you will need to press "y" whenever the application starts
+    "AcceptTermsOfService": true,
+
+    // You must at least one domain name
+    "DomainNames": [ "example.com" ],
+
+    // You must specify an email address to register with the certificate authority
+    "EmailAddress": "it-admin@example.com"
+  }
+}
```
Comment (MihaZupan) [inline docs/docfx/articles/lets-encrypt.md:51]:
```suggestion
{
  "Urls": "http://*:80;https://*:443",

  "Logging": { ... },

  "ReverseProxy": {
    "Routes": { ... },
    "Clusters": { ... }
  },

  "LettuceEncrypt": {
    // Set this to automatically accept the terms of service of your certificate authority.
    // If you don't set this in config, you will need to press "y" whenever the application starts
    "AcceptTermsOfService": true,

    // You must at least one domain name
    "DomainNames": [ "example.com" ],

    // You must specify an email address to register with the certificate authority
    "EmailAddress": "it-admin@example.com"
  }
}
```

I would keep the config example focused on `LettuceEncrypt`. We're already not including all the YARP boilerplate (clusters & routes), I think we can skip the rest too.

File: samples/ReverseProxy.LetsEncrypt.Sample/Properties/launchSettings.json [hunk 9]
```diff
@@ -0,0 +1,13 @@
+﻿{
+  "$schema": "http://json.schemastore.org/launchsettings.json",
+  "profiles": {
+    "YARP Proxy Sample": {
```
Comment (MihaZupan) [inline samples/ReverseProxy.LetsEncrypt.Sample/Properties/launchSettings.json:4]:
```suggestion
    "ReverseProxy.LetsEncrypt.Sample": {
```

File: samples/ReverseProxy.LetsEncrypt.Sample/appsettings.json [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        }
+      }
+    }
+  },
+  "LettuceEncrypt": {
+    // Set this to automatically accept the terms of service of your certificate authority.
+    // If you don't set this in config, you will need to press "y" whenever the application starts
+    "AcceptTermsOfService": true,
+
+    // You must at least one domain name
```
Comment (MihaZupan) [inline samples/ReverseProxy.LetsEncrypt.Sample/appsettings.json:44]:
```suggestion
    // You must specify at least one domain name
```

File: docs/docfx/articles/lets-encrypt.md (hunk 7 above)
Comment (greenEkatherine) [reply docs/docfx/articles/lets-encrypt.md:4]: Agree, will replace it

File: samples/ReverseProxy.LetsEncrypt.Sample/README.md [hunk 11]
```diff
@@ -0,0 +1,13 @@
+# Lets Encrypt Sample
+
+[Lets Encrypt](https://letsencrypt.org/) is a certificate authority (CA) that provides HTTPS (SSL/TLS) certificates for free. This sample shows how to add Lets Encrypt for TLS termination in YARP by integrating with [LettuceEncrypt](https://github.com/natemcmaster/LettuceEncrypt). It allows to set up TLS between the client and YARP and then use HTTP communication to the backend.
```
[the remark this answers is not in the document]
Comment (greenEkatherine) [reply samples/ReverseProxy.LetsEncrypt.Sample/README.md]: Replaced as well

File: docs/docfx/articles/lets-encrypt.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    public void ConfigureServices(IServiceCollection services)
+    {
+        services.AddLettuceEncrypt();
+    }
+}
+```
+
+For more options (i.e. saving certificates) see examples in [LettuceEncrypt doc](ttps://github.com/natemcmaster/LettuceEncrypt).
+
+## Middleware
```
Comment (Tratcher) [inline docs/docfx/articles/lets-encrypt.md:57]:
```suggestion
## Kestrel Endpoints
```

**Review by Tratcher — APPROVED** [review]

**Pull request merged**